src/dune_lang/syntax.ml

```diff
@@ -44,22 +44,79 @@ module Version = struct
 end
 
 module Supported_versions = struct
-  type t = int Int.Map.t
-
-  let to_dyn = Int.Map.to_dyn Int.to_dyn
-
-  let make = Int.Map.of_list_exn
-
-  let greatest_supported_version t = Option.value_exn (Int.Map.max_binding t)
-
-  let is_supported t (major, minor) =
+  (* The extension supported versions are declared using an explicit list of all
+     versions but stored as a map from major versions to maps from minor version
+     to dune_lang required versions *)
+  type t = Version.t Int.Map.t Int.Map.t
+
+  let to_dyn t = Int.Map.to_dyn (Int.Map.to_dyn Version.to_dyn) t
+
+  (* We convert the exposed extension version type: `(Version.t * [ `Since of
+     Version.t ]) list` which is a list of fully qualified versions paired with
+     the corresponding dune_lang version. To the internal representation:
+     `(Version.t Int.Map.t) Int.Map.t` which is a list of major versions paired
+     with lists of minor versions paires with a dune_lang version. *)
+  let make (versions : (Version.t * [ `Since of Version.t ]) list) : t =
+    let v =
+      List.fold_left versions
+        ~init:(Int.Map.empty : Version.t Int.Map.t Int.Map.t)
+        ~f:(fun major_map (v_ext, `Since v_lang) ->
+          let major, minor = v_ext in
+          let minor_map =
+            match Int.Map.find major_map major with
+            | Some m -> m
+            | None -> Int.Map.empty
+          in
+          let minor_map = Int.Map.add_exn minor_map minor v_lang in
+          Int.Map.update major_map major ~f:(function
+            | Some _minor_map -> Some minor_map
```

Comment:
I was a bit surprised that we were ignoring this value, but then realised that it is because we already fetched it a few lines above. What do you think of lifting the `update` around the `minor_map` updating? I feel like it would make the intent more explicit.

Reply:
I agree with you. Better https://github.com/ocaml/dune/pull/3270/commits/1cd331014c05124db45c378138ff035719db3e42 ?

Reply:
Definitely better!